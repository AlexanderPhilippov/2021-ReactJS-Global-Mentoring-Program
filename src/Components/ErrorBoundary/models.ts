import { ReactNode } from 'react'

export interface ErrorBoundaryProps {
    children: ReactNode
}

export interface ErrorBlockProps {
    error?: Error
}

export interface ErrorBoundaryState extends ErrorBlockProps {
    hasError: boolean
}
