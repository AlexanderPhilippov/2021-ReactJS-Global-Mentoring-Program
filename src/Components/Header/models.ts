export interface SearchPanelProps {
    modalToggle: React.Dispatch<React.SetStateAction<boolean>>
}

export interface MovieDetailsPanelProps {
    closeAction: () => void
}
