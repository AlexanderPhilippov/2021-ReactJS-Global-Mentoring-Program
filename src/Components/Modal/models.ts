export interface ModalProps {
    isOpen: boolean
    closeAction: () => void
}
