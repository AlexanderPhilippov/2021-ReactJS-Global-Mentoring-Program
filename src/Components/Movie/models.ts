export interface IMovieBase {
    title: string,
    tagline: string;
}