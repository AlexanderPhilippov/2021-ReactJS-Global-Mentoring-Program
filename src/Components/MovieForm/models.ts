import { MovieModel } from 'Components/Movie/models'

export enum MovieFormAction {
    ADD = 'Add movie',
    EDIT = 'Edit movie',
    DELETE = 'Delete movie',
}

export interface MovieFormProps {
    action?: MovieFormAction
    movie?: MovieModel
}

export interface FormikMovieModel
    extends Omit<Partial<MovieModel>, 'genres' | 'runtime' | 'vote_average'> {
    runtime: string
    genres: string
    vote_average: string
}

export interface FormBlockProps {
    touched?: boolean
    errorText?: string
    containerClassName: string
    name: string
    type?: string
    label: string
    placeholder?: string
    component?: string
}
