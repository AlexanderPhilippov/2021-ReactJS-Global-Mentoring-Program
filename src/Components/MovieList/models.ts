import { MovieModel } from 'Components/Movie/models'
import { MovieFormAction } from 'Components/MovieForm/models'

export interface MoviesResponseModel {
    data: MovieModel[]
    totalAmount: number
    offset: number
    limit: number
}

export interface MoviesHeaderProps {
    total?: number
}

export interface MovieListLocalState {
    isModalOpen: boolean
    selectedMovie?: MovieModel
    currentAction?: MovieFormAction
}

export interface MovieListReducerAction {
    payload: MovieListLocalState
}

export interface MovieListState extends Partial<MoviesResponseModel> {
    isLoading: boolean
    selectedGenre?: string
    sortBy?: string
    error?: string
}

export interface MovieListAction {
    type: string
    payload?: MoviesResponseModel
    error?: string
}
