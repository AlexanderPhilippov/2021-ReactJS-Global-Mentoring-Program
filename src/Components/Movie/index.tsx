import React from 'react'
import { useLocation, Link } from 'react-router-dom'
import { MovieProps } from './models'
import MovieCardMenu from './MovieCardMenu'
import './styles.scss'

const Movie: React.FC<MovieProps> = ({ movie }) => {
    const location = useLocation()
    const searchParams = new URLSearchParams(location.search)

    searchParams.set('movie', movie.id.toString())
    const linkTo = `${location.pathname}?${searchParams.toString()}`

    return (
        <Link className="movie-card" to={linkTo}>
            <MovieCardMenu movieId={movie.id} />
            <div className="movie-card__image">
                <img src={movie.poster_path} alt={`${movie.title} poster`} />
            </div>
            <div className="movie-card__title">{movie.title}</div>
            <div className="movie-card__release_date">
                {movie.release_date?.split('-')[0]}
            </div>
            <div className="movie-card__genres">{movie.genres.join(', ')}</div>
        </Link>
    )
}

export default Movie
