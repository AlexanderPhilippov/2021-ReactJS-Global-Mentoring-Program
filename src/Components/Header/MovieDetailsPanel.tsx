import React, { useContext, useEffect } from 'react'
import HeaderTextLogo from './HeaderTextLogo'
import MovieDetails from 'Components/MovieDetails'
import SearchIcon from 'Assets/Images/search.png'
import { Context } from 'Components/Context'
import { useSelector } from 'react-redux'
import { AppState } from 'src/Store/rootReducer'
import { getMovieByIdSelector } from 'Components/MovieList/selectors'
import { useHistory, useLocation } from 'react-router-dom'

const MovieDetailsPanel: React.FC = () => {
    const history = useHistory()
    const location = useLocation()
    const searchParams = new URLSearchParams(location.search)
    const { context, setContext } = useContext(Context)
    const handleClick = () => {
        searchParams.delete('movie')
        history.push({
            pathname: location.pathname,
            search: searchParams.toString(),
        })
        window.scrollTo({ top: context?.pageYOffset, behavior: 'smooth' })
        setContext()
    }

    const movie = useSelector((state: AppState) =>
        getMovieByIdSelector(state, context?.movieId)
    )

    useEffect(() => {
        if (!movie) setContext()
    }, [movie])

    return (
        <>
            <HeaderTextLogo />
            <img
                className="header__search-icon"
                src={SearchIcon}
                onClick={handleClick}
            />
            {movie && <MovieDetails {...movie} />}
        </>
    )
}

export default MovieDetailsPanel
