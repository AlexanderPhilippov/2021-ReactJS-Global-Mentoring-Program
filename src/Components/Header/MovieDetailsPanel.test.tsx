/**
 * @jest-environment jsdom
 */
import React from 'react'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import MovieDetailsPanel from './MovieDetailsPanel'
import { MemoryRouter } from 'react-router-dom'
import { MovieModel } from 'Components/Movie/models'

const movie: MovieModel = {
    id: 123,
    genres: ['action, dramma, comedy'],
    overview: 'Jest is tests',
    poster_path: 'http://test.images.com/poster/image',
    runtime: 180,
    title: 'Jest test movie object',
    release_date: '2021-11-07',
    vote_average: 10,
}

jest.mock('Components/MovieList/selectors', () => ({
    getMovieByIdSelector: () => movie,
}))

const mockStore = configureStore()
it('MovieDetailsPanel', () => {
    const store = mockStore({})
    const { asFragment } = render(
        <Provider store={store}>
            <MemoryRouter>
                <MovieDetailsPanel />
            </MemoryRouter>
        </Provider>
    )
    expect(asFragment()).toMatchSnapshot()
})
