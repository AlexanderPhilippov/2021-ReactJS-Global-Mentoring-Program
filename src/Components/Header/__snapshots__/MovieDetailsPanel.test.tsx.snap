// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetailsPanel 1`] = `
<DocumentFragment>
  <div
    class="header__text-logo"
  >
    <span>
      netflix
    </span>
    roulette
  </div>
  <a
    href="/"
  >
    <img
      class="header__search-icon"
      src=""
    />
  </a>
  <div
    class="movie-details"
  >
    <div
      class="movie-details__poster"
    >
      <img
        src="http://test.images.com/poster/image"
      />
    </div>
    <div
      class="movie-details__info"
    >
      <div
        class="movie-details__info-title"
      >
        Jest test movie object
      </div>
      <div
        class="movie-details__info-rating"
      >
        10
      </div>
      <div
        class="movie-details__info-genres"
      >
        action, dramma, comedy
      </div>
      <div
        class="movie-details__info-year"
      >
        2021
      </div>
      <div
        class="movie-details__info-runtime"
      >
        3hÂ 0min
      </div>
      <div
        class="movie-details__info-overview"
      >
        Jest is tests
      </div>
    </div>
  </div>
</DocumentFragment>
`;
